moveit_core/robot_state/src/conversions.cpp

```diff
@@ -284,12 +293,23 @@ static void _msgToAttachedBody(const Transforms* tf, const moveit_msgs::Attached
           }
         }
 
-        // transform poses to link frame
-        if (!Transforms::sameFrame(aco.object.header.frame_id, aco.link_name))
+        std::map<std::string, Eigen::Isometry3d> subframe_poses;
+        for (std::size_t i = 0; i < aco.object.subframe_poses.size(); ++i)
         {
+          Eigen::Isometry3d p;
+          tf2::fromMsg(aco.object.subframe_poses[i], p);
+          std::string name = aco.object.subframe_names[i];
+          subframe_poses[name] = p;
+        }
+
+        // Transform shape poses and subframes to link frame
+        if (!Transforms::sameFrame(aco.object.header.frame_id, aco.link_name))
+        {          
+          bool frame_found = false;
           Eigen::Isometry3d t0;
-          if (state.knowsFrameTransform(aco.object.header.frame_id))
-            t0 = state.getFrameTransform(aco.object.header.frame_id);
+          t0 = state.getFrameTransform(aco.object.header.frame_id, frame_found);
+          if (frame_found)
+            ;
```

Comment:
Prefer `if (not frame_found) { // put canTransform call here }`.

Reply:
Added, I don't like the if/else cascade without brackets anyway.